raiden/tests/unit/test_sqlite.py

```diff
@@ -0,0 +1,171 @@
+        token_network_address=factories.make_address(),
+        token=factories.make_address(),
+        pkey=factories.HOP1_KEY,
+        sender=factories.HOP1,
+    )
+
+    return signed_transfer
+
+
+def make_from_route_from_counter(counter):
+    from_channel = factories.make_channel(
+        partner_balance=next(counter),
+        partner_address=factories.HOP1,
+        token_address=factories.make_address(),
+        channel_identifier=next(counter),
+    )
+    from_route = factories.route_from_channel(from_channel)
+
+    expiration = factories.UNIT_REVEAL_TIMEOUT + 1
+
+    from_transfer = factories.make_signed_transfer_for(
+        channel_state=from_channel,
+        amount=1,
+        initiator=factories.make_address(),
+        target=factories.make_address(),
+        expiration=expiration,
+        secret=sha3(factories.make_secret(next(counter))),
+        identifier=next(counter),
+        nonce=1,
+        transferred_amount=0,
+        locked_amount=None,
+        pkey=factories.HOP1_KEY,
+        sender=factories.HOP1,
+    )
+    return from_route, from_transfer
+
+
+def test_get_latest_state_change_by_data_field():
+    """ All state changes which contain a balance proof must be found by when
+    querying the database.
+    """
+    serializer = JSONSerializer
+    storage = SQLiteStorage(':memory:', serializer)
+    counter = itertools.count()
+
+    lock_expired = ReceiveLockExpired(
+        balance_proof=make_signed_balance_proof_from_counter(counter),
+        secrethash=sha3(factories.make_secret(next(counter))),
+        message_identifier=next(counter),
+    )
+    transfer_direct = ReceiveTransferDirect(
+        token_network_identifier=factories.make_address(),
+        message_identifier=next(counter),
+        payment_identifier=next(counter),
+        balance_proof=make_signed_balance_proof_from_counter(counter),
+    )
+    unlock = ReceiveUnlock(
+        message_identifier=next(counter),
+        secret=sha3(factories.make_secret(next(counter))),
+        balance_proof=make_signed_balance_proof_from_counter(counter),
+    )
+    transfer_refund = ReceiveTransferRefund(
+        transfer=make_signed_transfer_from_counter(counter),
+        routes=list(),
+    )
+    transfer_refund_cancel_route = ReceiveTransferRefundCancelRoute(
+        routes=list(),
+        transfer=make_signed_transfer_from_counter(counter),
+        secret=sha3(factories.make_secret(next(counter))),
+    )
+    mediator_from_route, mediator_signed_transfer = make_from_route_from_counter(counter)
+    action_init_mediator = ActionInitMediator(
+        routes=list(),
+        from_route=mediator_from_route,
+        from_transfer=mediator_signed_transfer,
+    )
+    target_from_route, target_signed_transfer = make_from_route_from_counter(counter)
+    action_init_target = ActionInitTarget(
+        route=target_from_route,
+        transfer=target_signed_transfer,
+    )
+
+    statechange_balanceproof = [
```

Comment:
Note: This is quite bad, since new balance proof state changes are not automatically added. One alternative is to use a metaclass registry, which I honestly dislike, another alternative is to change how we shape our state changes, recommendations are welcome.